awx_collection/plugins/modules/tower_group.py

```diff
@@ -98,9 +112,21 @@ def main():
     description = module.params.get('description')
     state = module.params.pop('state')
     variables = module.params.get('variables')
+    hosts = module.params.get('hosts')
+    groups = module.params.get('groups')
 
     # Attempt to look up the related items the user specified (these will fail the module if not found)
     inventory_id = module.resolve_name_to_id('inventories', inventory)
+    hosts_ids = None
+    if hosts is not None:
+        hosts_ids = []
+        for item in hosts:
+            hosts_ids.append( module.resolve_name_to_id('hosts', item) )
```

Comment:
This is okay for some other modules, and as a first pass. But for the tower_group, using the contextual information of the inventory in the lookup is really needed for it to work without bugs. Otherwise, adding a child group of "aws" will almost certainly return duplicates when looking up only with name because multiple inventories will have it.

Reply:
For sure. I hit this myself with testing with 'localhost', it was finding localhost in another inventory. This is pretty common for anything that we lookup past the primary asset the module is dealing with. For now we accepted the pattern of just returning a message like "name was not unique enough, please pass an ID". I really think we need a tower_lookup module that would take a bunch of criteria and spit back an ID. Or maybe some of the work you talked about doing with the named URL might be another way to go.

Reply:
> I really think we need a tower_lookup module that would take a bunch of criteria and spit back an ID.

Totally different direction than anything we've done so far, but I agree.

Collections can host several types of content, and this is a lookup plugin if I understand it correctly. In broad terms, this is workable, and an amazing idea. I still anticipate a lot of difficulties in drafting requirements for it.

I believe tower-cli would accept an integer if `int()` worked. That's probably less likely to fly now that Ansible enforces strict types for parameters. But the only point I can imagine for a lookup plugin is to return the id.

Reply:
Right now its coded so that you can pass a name or an int. If it detects an int it will do the lookup by ID rather than name. 

The tower_lookup would be a good way to go from "human lookup criteria" to an int which the existing modules can consume. I think that module would be a better solution than having all of the modules coded up to take additional parameters to further clarify what things you are trying to lookup. i.e. if a module needs a credential it would also take fields like credential_type, credential_user, credential_team, credential_org, etc. This would explode the number of parameters for the modules.

Alternatively, if we were willing to drop backwards compatibility, we could make credential (and any lookup type fields) a dict that could contain any fields you wanted to try and perform the lookup.  i.e. 
```
credential:
    id: 7
```
or
```
credential:
    name: My Cred
    organization: My Org
```
This whole dict could be send to the lookup method and then it could just simply return the found item or "Unable to find the credential with data. Please be more specific".

Reply:
> I think that module would be a better solution

That plugin? Modules are 1 type of plugin. Lookup, inventory, docs are other types.

> if a module needs a credential it would also take fields like credential_type, credential_user, credential_team, credential_org, etc. This would explode the number of parameters for the modules.

There are formal published rules for the lookup fields. For credential it is `<name>++<credential_type.name>+<credential_type.kind>++<organization.name>`. Team, user, etc. is not part of that.

Should the `tower_credential` module take organization and credential_type params? Of course. Should it take a param for credential type kind? Unfortunately yes. There is no way around that in order to still be correct. If that is ugly, that's an issue what what the API has selected for these fields. Working around bad API behavior is a major part of what ultimately led to tower-cli getting deprecated.

----
Changing types from strings to dicts is not something I want to consider. Part of the reason is that the lookup plugin idea is truly a _good_ idea and does the same thing but better.

The lookup plugin can allow people writing playbooks to look things up in ways that we have not anticipated, as long as it passes arbitrary stuff into the query parameters.

> Right now its coded so that you can pass a name or an int. If it detects an int it will do the lookup by ID rather than name.

That's good, and works for the purpose of the lookup plugin. We should have test coverage for it.

Reply:
Ah, yes I wasn't even thinking of a lookup plugin, that is even more of a fantastic idea! then playooks could do something like:
```
- name: Build a job template
  tower_job_template:
    name: "{{ lookup('tower_asset', <criteria>) }}"
    credential: "{{ lookup('tower_asset', <credential criteria>) }}"
```